feat: update blockcapturer to capture outcomes

DavePearce opened the pull request:
Update `BlockCapturer` to capture outcomes.  These are then checked (when present) as part of a `replayTest()`.  The benefit of this is that it helps to ensure the testing environment reflects mainnet.

DavePearce committed:
Enable capture & checking of transaction outcomes

This updates the `BlockCapturer` to capture details related to
transaction outcomes.  Specifically, it records whether or not the
transaction suceeded, what output data was generated, what logs were
generated, how much gas was used and what accounts were self destructed.

The expected outcome is checked against the actual outcome, and a
suitable exception raised if there is a discrepency.  However, if the
outcome is missing (i.e. because its an old trace which pre-dates this
feature) then a log is reported that the outcome "was not checked".

Finally, it should be noted that we could attempt to capture more
information about what happened during transaction executing.
Specifically, what the state differences were.

DavePearce committed:
Record touched state in `TransactionResultSnapshot`

This updates the `BlockCapturer` to record the set of accounts and
storage locations which were touched by a given transaction.  This
includes those which were modified, but also those which were just read.

This allows the testing mechanism to ensure that the state is exactly as
we expected after each transaction.

DavePearce committed:
Fix Transaction Result Checking

There was a problem with the position at which result checking was being
performed, in that it didn't exactly match the position where the result
snapshot data was being generated.  This is resolved using a wrapper
around ZkTracer which handles checking of transaction results at the
correct position.

DavePearce committed:
Fix `ToyWorld.commit()`

This puts through a simple fix for `ToyWorld.commit()` which became
evident from block `2247098`.  Since this matches the code in
`SimpleWorld`, it seems to me there is a bug there.

FlorianHuc commented on testing/src/main/java/net/consensys/linea/testing/ToyExecutionEnvironment.java:
```diff
@@ -211,7 +223,7 @@ private void execute() {
 
       this.testValidator.accept(result);
-      this.zkTracerValidator.accept(tracer);
+      // this.zkTracerValidator.accept(tracer);
```
if it's not used, should we delete it?

DavePearce replied:
That's a mistake.  It should not have been commented out.

DavePearce replied:
That's a mistake.  It should not have been commented out.

however, in fact, the enclosing `execute()` method is no longer used anywhere as all the existing tests now use `ToyExecutionEnvironmentV2`.  So we could actually drop that, and I would suggest doing so (perhaps in a separate PR though).

FlorianHuc commented on arithmetization/src/main/java/net/consensys/linea/blockcapture/reapers/Reaper.java:
```diff
@@ -39,34 +43,79 @@
+    txSnapshot.setTransactionResult(resultSnapshot);
+    // Reset for next transaction
+    txIndex++;
+    this.txStorage = new StorageReaper();
```
Given we initialize them in the enterTransaction, I we shouldn't need to initialize them here. Should we instead add a close method here, and in the enter check that the reapers are either null or close? This way we would be sure that tx and entered then exited one after the other?

DavePearce replied:
Originally I did something like that, but I had a problem with serialisation.  Let me revisit that.

DavePearce replied:
Ok, have resolved the problem here.  It was that, for some transactions, `traceStartTransaction()` is never called!  Instead, `tracePrepareTransaction()` is always called but we can go straight to `traceEndTransaction()` from there without calling `traceStartTransaction()`.

I've also refactored the way in which the `TransactionChecker` was integrated.  Previously, it was a wrapper around an existing `OperationTracer` (which was `ZkTracer`).  However, now its a standalone implementation and I use a separate mechanism for multiplexing tracers together (defined in `ConflationAwareOperationTracer`).  This has an additional benefit that I can relatively easily include the `BlockCapturer` for testing purposes.

DavePearce committed:
Refactor for transactions which are non-starters

This puts in place a refactoring which allows `txStorage` and
`txAddresses` to be initialised with `null`.  A key challenge was that
not all transactions go through `traceStartTransaction()`.  Instead,
some transactions go `tracePrepareTransaction()` and then straight to
`traceEndTransaction()`.

In addition, this refactors `TransactionChecker` so that it no longer
wraps an existing tracer.  Instead, we can "sequence" tracers using
`ConflationAwareOperationTracer.sequence()` making it relatively easy to
combine two tracers together.  Furthermore, this helps with testing the
`BlockCapturer` as it becomes easier to run it from
`ToyExecutionEnvironment`.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/blockcapture/BlockCapturer.java:
```diff
@@ -68,22 +68,27 @@ public void traceStartBlock(BlockHeader blockHeader, BlockBody blockBody) {
+  }
+
   @Override
   public void traceStartTransaction(WorldView worldView, Transaction transaction) {
```
we should plug only on `tracePrepareTransaction` besu hook. `traceStartTransaction` is just after, when the upfront gascost (tx + dtat + access list + deployment ) is paid by the sender.

DavePearce replied:
We need to use `tracePrepareTransaction()` here.  For some transactions, `traceStartTransaction()` is never called and execution goes straight to `traceEndTransaction()`.  I'm not sure exactly the circumstances, but its either for an invalid transaction or perhaps one that has no code associated with it.  Eitherway, without this, we will get a `NullPointerException` in `traceEndTransaction()` as `txStorage` and `txAddresses` will be `null`.

DavePearce replied:
I only found this out from testing, tbh.

DavePearce replied:
Anyway, sorry, I guess the point is: I stuck with `traceStartTransation()` because that is what the `BlockCapturer` was originally doing.  However, yes, am happy to just merge into `tracePrepareTransaction()`.

letypequividelespoubelles replied:
yes, we should be plugged only at `tracePrepareTransaction()` and remove `traceStartTransaction()`

DavePearce replied:
Yeah sorry for some reason I read you original message the other way around.  Have made the change and just waiting for tests to go green.

DavePearce committed:
Drop `Reaper.traceStartTransaction()`.

letypequividelespoubelles reviewed (approved):
LGTM, TY
